---
title: "Class 6: R functions"
author: "Kyle Wittkop (A18592410)"
format: pdf
toc : true 
---

## Backround 

All functions in R have at least 3 things : 

- A **name** that we use to call the function

- Has one or more input **arguments** 

- The **body**: the lines of R code that do the work 

## Our first function 

Lets write a silly wee function called `add()` to add to add some numbers (the input function)

```{r}
add <- function(x,y) {
  x + y 
} 
```

Now we can use this function 

```{r}
add(100,1)
```
```{r}
add(x=10,y=10)
add( x=c(100,1,100), y=(1))
add(x=c(100,1), y=c(1,100))
```
What if i give a multiple element vector to x and y 
```{r}
add(x=c(100,1), y=c(1,100))
```

Q. what if i give three inputs to the function 
```{r}
#add(x=c(100,1,y=1,z=1))
```

Q. What if i give only 1 input to the add function? 
```{r}
#add(x=100)
addnew <-function(x,y=1){
  x+y
}
```

```{r}
addnew(x=100)
addnew(x=100,y=200)
```
If we write out function with input arguments having no default value than the user will be required to set them when they use the function. we can give our input arguments defualt values by setting them equal to some sensible value e.g. y=1 

## A second function 

Lets make something more interesting: make a sequence generating tool... 

The `sample()` function can be a useful starting point 

```{r}
sample(1:10, size=4)

```
>Q. Generate 9 random numbers from the input vector from 1 to 10 

```{r}
sample(1:10, size=9)
```
>Q. Generate 12 random numbers from the input vector from 1 to 10? 

```{r}
sample(1:10, size =12, replace=TRUE)
```

>Q. Write code for the sample function that generates nucleotide sequences of length 6?

```{r}
sample(c("A","T","C","G"), size = 6, replace = TRUE)



```

>Q. Write a first function `generate_DNA()` that returns a user specified length DNA sequence: 

```{r}

generate_DNA <-function(x=6) {
 sample(c("A","T","C","G"), size=x, replace = TRUE)
}
```

```{r}
generate_DNA()
generate_DNA(100)
generate_DNA(10)
```

**KEY POINTS**

Every function in R looks fundamentally the same in terms of its structure, Basically 3 things : name, input, body 

```
name <- function(input) {
  Body 
}
``` 

Functions can have multiple inputs. These can be required arguments or "optional" with optional arguments having a set of default values.

> Q. Modify and improve our generate_DNA function to return its generated sequence in a more standard format like "ATGACA" rather than the vector "A","C","G","A" 

```{r}
generate_DNA <-function(x=6, fasta=TRUE) {
 ans <- sample(c("A","T","C","G"), 
        size=x,  replace = TRUE)
 if(fasta)
   cat("Single-element vector output ")
ans <- paste(ans, collapse="")
  return(ans)
}

generate_DNA(fasta=FALSE)

```

the `paste()` function, its job is to join up or stick together (a.k.a paste) input strings together 

```{r}
paste(c("alice","barry"), "love R", sep = " Does not ")
?paste

```

Flow control means where the R brain goes in your code 
```{r}
good_mood <- FALSE

if(good_mood) {
  cat("Great!")
} else{ 
  cat("Bummer")
}
```

## A protein generating function 

> Q. Write a function that generates a user specified length protein sequence.

```{r}
Generate_Prot <-function(x) {
 ans <- sample(c("A","R","N","D","C","Q","E","G","H","I","L","K","M","F","P","S","T","W","Y","V"), 
        size=x,  replace = TRUE)
 ans <- paste(ans,collapse="" )
  return(ans)
}
Generate_Prot(6)
```


>Q. Use that function to generate random protien sequences between 6 and 12 

```{r}
Generate_Prot(6)
Generate_Prot(12)
```
```{r}
for(i in 6:12) {
  cat(">",i, sep="", "\n")
  cat(Generate_Prot(i), "\n")
}
```


>Q. are any of your sequences unique i.e not found in nature 

Yes, sequences that were 8 amino acids or greater were unique in nature. 
